app/controllers/device_controller.rb

```diff
@@ -1,5 +1,59 @@
 class DeviceController < ApplicationController
+  before_action :authorize_request, :only=>[:add_tag, :delete_tag]
+
   def show
     @devices = Device.all
   end
+
+  def raw_data
```

Comment:
This `raw-data` implementation was migrated from the previous `ReportController#raw_data`, where the algorithm of calculating the max emission per core is incorrect. A possible correction is provided below:
```
def raw_data
  devices = Device.left_joins(:user)
  .select('devices.*, users.username, max / (cpus * cores_per_cpu) AS max_per_core')
  .order(:max_per_core)
  response = {}.tap do |res|
    max_device = devices.last
    res[:max_main] = max_device.max_per_core.round(3);
    res[:header] = {}.tap do |h|
      h[:user] = 'User'
      h[:platform] = 'Platform'
      h[:location] = 'Location'
      h[:core_number] = 'No. cores'
      h[:ram] = 'RAM (GB)'
      h[:main] = 'Carbon emissions per core at full load (kgCO2eq/h)'
    end
    rank = 1
    res[:devices] = devices
    .group_by(&:max_per_core)
    .values.flat_map do |dev_group|
      current_rank = rank
      rank += dev_group.size
      dev_group.map do |dev|
        {}.tap do |new_dev|
          new_dev[:rank] = current_rank
          new_dev[:user] = dev.username || 'Anonymous'
          new_dev[:platform] = dev.platform
          new_dev[:location] = dev.location
          new_dev[:core_number] = dev.cpus * dev.cores_per_cpu
          new_dev[:ram] = dev.ram_units * dev.ram_capacity_per_unit
          new_dev[:main] = dev.max_per_core.round(3);
        end
      end
    end
  end
  render json: response
end
```

Reply:
Could you explain what `res[:max_main]` is used for here? It's causing an issue in that it can't be calculated if there aren't any devices, because `max_device` becomes `nil` and so doesn't have a `max_per_core` attribute.

If I set `res[:max_main]` to `nil` in the case that there are no devices at all, would that be appropriate? It seems to fix the issue but just want to clarify what that value is actually doing.

Reply:
![image](https://github.com/openflighthpc/carbon-leaderboard/assets/98431935/d97d0b67-8a0f-4377-94c9-b4ad349baf2e)
Yes it could be nil or 0 if there are no devices. This is used for calculating the length of the glowing bar.